### PR TITLE
Write Tests for *-Advice Cmdlets #278

### LINKED ISSUE #278: Write Tests for *-Advice Cmdlets
# Describe "Functionality"

The `*-Advice` public commands do not have any tests verifying their behaviour.

## Context "Private or Public? What are the use cases? Parameters? Options?"

`Show-Advice` and `Register-Advice` are public module functions.
Both are completely lacking any tests as of yet.
Refer to each command's documentation and working code to determine appropriate tests.

### PR DESCRIPTION
**Pull request by manoj2994**
﻿# PR Summary
Created Tests For Show-Advice and Register-Advice with the possibilities of chances for execution and made a change on the Register-Advice function where Select-String '-Path' will only accept pipeline by property name.Previously a path variable was passed to the pipeline which is not taking by the Select-String

Fixes #278 
Resolves #278 

## Changes

- Changed the Reigister-Advice Function Select-String Cmdlet Execution.
- Added Register-Advice Tests File in the Public Folder under Tests
- Added Show-Advice Tests File in the Public Folder under Tests

### CODE REVIEW

File: Tests/Functions/Public/Show-Advice.Tests.ps1 [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+            It "Write-ConsoleLine Without Paramter Should be Called " {
+                Show-Advice -Name "Profile" | Should -BeNullOrEmpty
+            }
+
+        }
+
+    }
+
+
+
```
Comment (vexx32) [inline Tests/Functions/Public/Show-Advice.Tests.ps1]: Please trim out blank lines at the end of a scriptblock, here and in the rest of the file(s). 

Thanks! :)

File: Tests/Functions/Public/Show-Advice.Tests.ps1 [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+                Show-Advice -Name "Profile" | Should -BeNullOrEmpty
+            }
+
+        }
+
+    }
+
+
+    
+}
```
[the remark this answers is not in the document]
Comment (manoj2994) [reply Tests/Functions/Public/Show-Advice.Tests.ps1]: Done

File: Tests/Functions/Public/Register-Advice.Tests.ps1 [hunk 3]
```diff
@@ -0,0 +1,59 @@
+#region Header
+if (-not (Get-Module PSKoans)) {
+    $moduleBase = Join-Path -Path $psscriptroot.Substring(0, $psscriptroot.IndexOf('\Tests')) -ChildPath 'PSKoans'
+    Import-Module $moduleBase -Force
+}
+#. "C:\Git\PSKoans\PSKoans\Public\Register-Advice.ps1"
+#endregion
```
[the remark this answers is not in the document]
Comment (manoj2994) [reply Tests/Functions/Public/Register-Advice.Tests.ps1]: Thanks for reviewing it ,ill make the changes

File: Tests/Functions/Public/Register-Advice.Tests.ps1 [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+        $testdata = @{1='AllUsersAllHosts'},@{1= 'AllUsersCurrentHost'},@{1='CurrentUserAllHosts'},@{1='CurrentUserCurrentHost'}
+        It "Checking with the possible Parameter Values" -TestCases $testdata{
+            param($1)
+            Register-Advice $1 | should -BeNullOrEmpty 
+            
+        }
+
+    }
+
+}
```
[the remark this answers is not in the document]
Comment (manoj2994) [reply Tests/Functions/Public/Register-Advice.Tests.ps1]: Formatted the indented blocks

File: Tests/Functions/Public/Show-Advice.Tests.ps1 [hunk 5]
```diff
@@ -0,0 +1,41 @@
+#region Header
+if (-not (Get-Module PSKoans)) {
+    $moduleBase = Join-Path -Path $psscriptroot.Substring(0, $psscriptroot.IndexOf('\Tests')) -ChildPath 'PSKoans'
+    Import-Module $moduleBase -Force
+}
+#endregion
```
[the remark this answers is not in the document]
Comment (manoj2994) [reply Tests/Functions/Public/Show-Advice.Tests.ps1]: Changed as suggested

File: Tests/Functions/Public/Show-Advice.Tests.ps1 [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+            }
+            
+            It "Show-Advice Output Should be Void"{
+                $result | Should -BeNullOrEmpty
+            }
+        }
+
+        Context "Checking the Behaviour of  Function Calling with parameter" {
+
+            It "Write-ConsoleLine Without Paramter Should be Called " {
```
[the remark this answers is not in the document]
Comment (manoj2994) [reply Tests/Functions/Public/Show-Advice.Tests.ps1]: created tests and added a exception messae tests

File: Tests/Functions/Public/Show-Advice.Tests.ps1 [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+
+            It "should call Write-ConsoleLine with normal parameters" {
+                Show-Advice -Name "Profile.Advice" 
+                Assert-MockCalled -CommandName Write-ConsoleLine -ParameterFilter { $null -ne $Title }
+            }
+            It "should call Write-ConsoleLine without parameters" {
+                Show-Advice -Name "Profile.Advice" 
+                Assert-MockCalled -CommandName Write-ConsoleLine -ParameterFilter { $null -eq $Title }
+            }
+            It "Should Throw an Expected Messae if invalid parameter is passed" {
```
Comment (vexx32) [inline Tests/Functions/Public/Show-Advice.Tests.ps1]:
```suggestion
            It "should throw an error if the requested file cannot be found" {
```

File: Tests/Functions/Public/Show-Advice.Tests.ps1 [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+        }
+
+        Context "Behaviour with -Name Parameter" {
+            
+            Mock Write-ConsoleLine { }
+
+            It "should call Write-ConsoleLine with normal parameters" {
+                Show-Advice -Name "Profile.Advice" 
+                Assert-MockCalled -CommandName Write-ConsoleLine -ParameterFilter { $null -ne $Title }
+            }
```
Comment (vexx32) [inline Tests/Functions/Public/Show-Advice.Tests.ps1]:
```suggestion
            }

```

File: Tests/Functions/Public/Show-Advice.Tests.ps1 [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+            It "should call Write-ConsoleLine without parameters" {
+                Show-Advice -Name "Profile.Advice" 
+                Assert-MockCalled -CommandName Write-ConsoleLine -ParameterFilter { $null -eq $Title }
+            }
```
Comment (vexx32) [inline Tests/Functions/Public/Show-Advice.Tests.ps1]:
```suggestion
            }

```

File: Tests/Functions/Public/Show-Advice.Tests.ps1 [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+                $result | Should -BeNullOrEmpty
+            }
+        }
+
+        Context "Behaviour with -Name Parameter" {
+            
+            Mock Write-ConsoleLine { }
+
+            It "should call Write-ConsoleLine with normal parameters" {
+                Show-Advice -Name "Profile.Advice"
```
Comment (vexx32) [inline Tests/Functions/Public/Show-Advice.Tests.ps1]:
```suggestion
                Show-Advice -Name "Profile" 
```
Oops, missed this one. This is what's causing the test failures currently, I think.

Comment (manoj2994) [reply Tests/Functions/Public/Show-Advice.Tests.ps1]: when im running tests on the environment it fails with just the name "Profile" but calling the function with the format like this "Profile.Advice".Ill commit these now and do check

File: Tests/Functions/Public/Register-Advice.Tests.ps1 [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+
+InModuleScope 'PSKoans' {
+    Describe "Register-Advice" {
+    
+        Context "Checkin the Behaviour if ProfileFolder and ProfilePath Not exists" {
+        
+            Mock Test-Path { $false } -ParameterFilter { $Path -eq $ProfileFolder }
+            Mock New-Item { }
+            Mock Test-Path { $false } -ParameterFilter { $Path -eq $ProfilePath }
+            Mock Set-Content { } -ParameterFilter { $Value -eq "Show-Advice" }
```
[the remark this answers is not in the document]
Comment (manoj2994) [reply Tests/Functions/Public/Register-Advice.Tests.ps1]: Done have put those mocks in before all blocks

**Pull request merged**